### ISSUE TITLE
Ivy cannot properly resolve ES v5.0.0 dependencies

### ISSUE THREAD

**Issue by ywelsch** [elastic/elasticsearch#21170]
The Maven pom definitions generated for ES v5.0.0+ use \* excludes to prevent transitive dependencies. This breaks compatibility with the dependency manager Apache Ivy as it incorrectly translates POMs with * excludes to Ivy XML with * excludes which results in the main artifact being excluded as well (see https://issues.apache.org/jira/browse/IVY-1531). This subsequently breaks the consumption of these dependencies from SBT (has a fix for 0.13.3 though), Grape, and other software that uses Ivy to resolve dependencies.

In Ivy the \* excludes of the form

```
<dependencies>
    <dependency>
      <groupId>io.netty</groupId>
      <artifactId>netty-buffer</artifactId>
      <version>4.1.5.Final</version>
      <scope>compile</scope>
      <exclusions>
        <exclusion>
          <groupId>*</groupId>
          <artifactId>*</artifactId>
        </exclusion>
      </exclusions>
    </dependency>
   ...
```

exclude the artifacts of `netty-buffer` itself as well.

This can be shown by running

```
ivy -confs default -dependency org.elasticsearch.plugin transport-netty4-client 5.0.0 -retrieve "[artifact](-[classifier]).[ext]"
```

which yields

```
:: loading settings :: url = jar:file:/usr/local/Cellar/ivy/2.4.0/libexec/ivy-2.4.0.jar!/org/apache/ivy/core/settings/ivysettings.xml
:: resolving dependencies :: org.elasticsearch.plugin#transport-netty4-client-caller;working
    confs: [default]
    found org.elasticsearch.plugin#transport-netty4-client;5.0.0 in public
    found io.netty#netty-buffer;4.1.5.Final in public
    found io.netty#netty-codec;4.1.5.Final in public
    found io.netty#netty-codec-http;4.1.5.Final in public
    found io.netty#netty-common;4.1.5.Final in public
    found io.netty#netty-handler;4.1.5.Final in public
    found io.netty#netty-resolver;4.1.5.Final in public
    found io.netty#netty-transport;4.1.5.Final in public
:: resolution report :: resolve 564ms :: artifacts dl 6ms
    ---------------------------------------------------------------------
    |                  |            modules            ||   artifacts   |
    |       conf       | number| search|dwnlded|evicted|| number|dwnlded|
    ---------------------------------------------------------------------
    |      default     |   8   |   0   |   0   |   0   ||   2   |   0   |
    ---------------------------------------------------------------------
:: retrieving :: org.elasticsearch.plugin#transport-netty4-client-caller
    confs: [default]
    2 artifacts copied, 0 already retrieved (740kB/23ms)
```

i.e. it just yields the following 2 artifacts:

```
$ ls
netty-common.jar
transport-netty4-client.jar
```

missing `netty-buffer` and others which use the \* excludes.

**Comment by retemaadi** [comment]
I have the same issue with "java.lang.NoClassDefFoundError: org/apache/logging/log4j/Logger"

**Comment by fabriziofortino** [comment]
@retemaadi I think yours is a different issue. With ES 5.0.0 you need to add logging dependencies. Have a look here https://discuss.elastic.co/t/issue-with-elastic-search-5-0-0-noclassdeffounderror-org-apache-logging-log4j-logger/64262/1

**Comment by retemaadi** [comment]
@fabriziofortino Thanks for the link!
However, after I added log4j, still I get class not found problem for Netty.

**Comment by ywelsch** [comment]
@retemaadi which build system are you using? Can you provide more details? Please create a new topic on https://discuss.elastic.co so that we can discuss your specific case there.

**Comment by jasontedor** [comment]
Supersedes #20926

**Comment by nik9000** [comment]
I'd really like to get this into 5.0.1. Which means someone's gotta claim it soon!

**Comment by ywelsch** [comment]
I've created #21234 as possible fix. Please let me know what you think @nik9000 @rjernst.

**Comment by retemaadi** [comment]
@ywelsch I'm using SBT in a Scala project and I tried to upgrade from ("org.elasticsearch" % "elasticsearch" % "2.4.0") to ("org.elasticsearch.client" % "transport" % "5.0.0"), which resulted in the exact failure that your described here.

I'm pretty sure, my issue will be resolved when this issue get fixed, do you still want me to create a new discussion topic?

**Comment by ywelsch** [comment]
@retemaadi Upgrading SBT to version 0.13.13 will also fix this (see https://github.com/sbt/sbt/pull/2731). No need to create a new discussion topic.

**Comment by retemaadi** [comment]
@ywelsch Thanks for the tip! I'll give it a try.

**Comment by retemaadi** [comment]
@ywelsch I tried it out with SBT 0.13.13 and still I get the error. Here is the stacktrace:

java.lang.NoClassDefFoundError: io/netty/channel/RecvByteBufAllocator
        at org.elasticsearch.transport.Netty4Plugin.getSettings(Netty4Plugin.java:39)
        at org.elasticsearch.plugins.PluginsService.lambda$getPluginSettings$0(PluginsService.java:85)
        at java.util.stream.ReferencePipeline$7$1.accept(ReferencePipeline.java:267)
        at java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1374)
        at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:481)
        at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:471)
        at java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:708)
        at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
        at java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:499)
        at org.elasticsearch.plugins.PluginsService.getPluginSettings(PluginsService.java:85)
        at org.elasticsearch.client.transport.TransportClient.buildTemplate(TransportClient.java:115)
        at org.elasticsearch.client.transport.TransportClient.<init>(TransportClient.java:228)
        at org.elasticsearch.transport.client.PreBuiltTransportClient.<init>(PreBuiltTransportClient.java:69)
        at org.elasticsearch.transport.client.PreBuiltTransportClient.<init>(PreBuiltTransportClient.java:65)

**Comment by ywelsch** [comment]
@retemaadi you probably need to clean your ivy cache first if you previously tried with an older version of SBT: `rm -r ~/.ivy2/cache/*`

**Comment by retemaadi** [comment]
@ywelsch Thanks for your time, but still I see the same problem. So, I gave up! I'm going to wait for 5.0.1

**Comment by ywelsch** [comment]
@retemaadi I tried it and it works for me with SBT 0.13.13.

My `build.sbt` build file:

```
name := "sbt-transport-example"

version := "1.0"

scalaVersion := "2.12.0"

libraryDependencies ++= Seq(
  "org.elasticsearch" % "elasticsearch" % "5.0.0",
  "org.elasticsearch.client" % "transport" % "5.0.0",
  "org.apache.logging.log4j" % "log4j-api" % "2.7",
  "org.apache.logging.log4j" % "log4j-core" % "2.7"
)
```

My Java class in `src/main/java`:

```
import org.elasticsearch.client.transport.TransportClient;
import org.elasticsearch.common.settings.Settings;
import org.elasticsearch.common.transport.InetSocketTransportAddress;
import org.elasticsearch.transport.client.PreBuiltTransportClient;

import java.net.InetAddress;
import java.net.UnknownHostException;

public class Main {
    public static void main(String... args) throws UnknownHostException {
        TransportClient client = new PreBuiltTransportClient(Settings.EMPTY)
                .addTransportAddress(new InetSocketTransportAddress(InetAddress.getByName("localhost"), 9300));
        System.out.println(client.admin().cluster().prepareState().get().getState().prettyPrint());
        client.close();
    }
}
```

If you cannot get it to work with SBT 0.13.13, please open an issue on `https://discuss.elastic.co` with additional information and ping me there (@ywelsch).

**Comment by ywelsch** [comment]
Closed by #21234

**Comment by retemaadi** [comment]
@ywelsch Thanks! It was the issue of bundled IntelliJ SBT.
